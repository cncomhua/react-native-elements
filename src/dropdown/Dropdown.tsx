import React from 'react';
import {
  View,
  Text,
  StyleSheet,
  StyleProp,
  ViewStyle,
  Dimensions,
  TouchableOpacityProps,
  TouchableOpacity,
  Animated,
  TextStyle,
  Easing,
  ScrollViewProps
} from 'react-native';
import Icon, { IconNode } from '../icons/Icon';
const {width,height} = Dimensions.get('screen')

export type ItemProps = {
    element:string |React.ReactElement<{}>;
    value:string|number|boolean;
    onPress?:Function;
    icon?:IconNode;
}
export type DividerStyle = {
    width?:number;
    endRaius?:number;
    color?:string;
}

export type DropdownProps = TouchableOpacityProps & ScrollViewProps & {
    title?: string |React.ReactElement<{}> ;
    titleStyle?:StyleProp<TextStyle>;
    titleContainerStyle?:StyleProp<ViewStyle>;
    chevron?:boolean;
    containerStyle?:StyleProp<ViewStyle>;
    chevronColor?:string;
    textStyle?: StyleProp<TextStyle>;
    items:Array<ItemProps>;
    maxHeight?:number;
    animationDuration?:number;
    onChange?:(value)=>void;
    icon?:IconNode;
    iconOpen?:IconNode;
    iconClose?:IconNode;
    textContainerStyle?:StyleProp<ViewStyle>;
    chevronSize?:number;
    disabled?:boolean;
    divider?:boolean;
    dividerStyle?:DividerStyle;
    theme?:'default'|'dark';
}
const Dropdown:React.FunctionComponent<DropdownProps> = (props: DropdownProps)=>{
    const {
        title,
        titleStyle,
        containerStyle,
        textContainerStyle,
        chevron=true,
        chevronColor,
        textStyle,
        items,
        onChange=(value)=>console.log('Please attach a method to this component'),
        maxHeight=height/10,
        animationDuration=200,
        icon,
        iconOpen,
        iconClose,
        chevronSize=20,
        disabled=false,
        dividerStyle={width:1},
        divider=true,
        theme='default',
        titleContainerStyle,
        ...attributes
    }=props;
    const slider = React.useRef<Animated.Value>(new Animated.Value(0)).current;
    const [sliderState,setSliderState] = React.useState<boolean>(false);
    const [currentPtr,setCurrentPtr] = React.useState<number>(-1);
    const extract = () => {
        slider.setValue(0);
        Animated.timing(slider, {
            toValue: 1,
            duration: animationDuration,
            easing:Easing.cubic,
            useNativeDriver:false,
        }).start();
    };
    const retreact = () => {
        slider.setValue(1);
        Animated.timing(slider, {
          toValue: 0,
          duration: animationDuration,
          easing:Easing.cubic,
          useNativeDriver:false,
        }).start(()=>setSliderState(false));
    };
    const sliderInterpolate = slider.interpolate({
        inputRange:[0,0.2,0.4,0.6,0.8,1],
        outputRange:[0*maxHeight/5,1*maxHeight/5,2*maxHeight/5,3*maxHeight/5,4*maxHeight/5,5*maxHeight/5]
    })
    return(
        <View style={
            StyleSheet.flatten([
                {
                    opacity:disabled?0.5:1,
                    zIndex:1,
                    borderWidth:1,
                    borderColor:theme==='default'?'black':"white"
                },
                containerStyle
            ])
            }>
            <TouchableOpacity {...attributes} 
                disabled={disabled}
                onPress={()=>{
                    if(sliderState){
                        retreact();
                    }
                    else{
                        extract();
                        setSliderState(true)
                    }
                }}
                style={StyleSheet.flatten([
                    {
                        justifyContent:"space-between",
                        flexDirection:'row',
                        overflow:"hidden",
                        borderWidth:0,
                        borderBottomWidth:sliderState?dividerStyle.width?dividerStyle.width:1:0,
                        borderColor:dividerStyle.color?dividerStyle.color:theme==='default'?'black':'white',
                        borderRadius:dividerStyle.endRaius?dividerStyle.endRaius:0,
                    },
                    titleContainerStyle
                ])}>
                <View  style={{alignSelf:'center'}}>
                    {currentPtr===-1?
                        typeof title==='string'?
                            <Text style={StyleSheet.flatten([
                                styles.defaultText,
                                {
                                    color:theme==='default'?'black':'white',
                                },
                                titleStyle,
                            ])}>{title}</Text>
                            :
                            title
                        :
                        typeof items[currentPtr].element==='string'?
                            <Text  style={StyleSheet.flatten([
                                styles.defaultText,
                                {
                                    color:theme==='default'?'black':'white'
                                },
                                titleStyle,
                                textStyle,
                            ])}>{items[currentPtr].element}</Text>
                            :
                            items[currentPtr].element
                    }
                </View>
                <View  style={{alignSelf:'center'}}>
                    {chevron?
                        icon?
                            icon
                            :
                            iconOpen?
                                sliderState?
                                    iconClose?iconClose:iconOpen:iconOpen
                                :
                                <Icon color={chevronColor?chevronColor:theme==='default'?'black':'white'} size={chevronSize} name={sliderState?"chevron-up":"chevron-down"} type="feather" />
                        :
                        null
                    }
                 </View>
            </TouchableOpacity>
            <Animated.ScrollView style={StyleSheet.flatten([
                    {
                        overflow:'hidden',
                        height:sliderInterpolate,
                    },

                ])}
                {...attributes}
                >
                {items.map((data,index)=>{
                    return(
                        <TouchableOpacity 
                            key={index} 
                            style={StyleSheet.flatten([
                                {
                                    flexDirection:'row',
                                    justifyContent:'space-between'
                                },
                                textContainerStyle
                            ])}
                            onPress={()=>{
                                retreact();
                                data.onPress?data.onPress():null;
                                onChange(data.value);
                                setCurrentPtr(index);
                            }}>
                            <Text  style={StyleSheet.flatten([
                                styles.defaultText,
                                {
                                    color:theme==='default'?'black':'white'
                                },
                                textStyle,
                            ])}>{data.element}</Text>
                            {data.icon?data.icon:null}
                        </TouchableOpacity>
                    )
                })}
            </Animated.ScrollView>
            </View>
    )
}

const styles = StyleSheet.create({
    defaultText:{
        color:'black',
        fontSize:15,
        paddingVertical:2,
        paddingHorizontal:2,
    }
})
export {Dropdown};
