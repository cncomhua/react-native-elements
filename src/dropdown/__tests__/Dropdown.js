import React from 'react';
import { Dropdown } from '../Dropdown';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { TouchableOpacity } from 'react-native';

describe('BottomSheet Component', () => {
  it('renders correctly', () => {
    const tree = renderer
      .create(
          <Dropdown 
            title="--selection--" 
            items={[
                {element:'Facebook',value:'Facebook'},
                {element:'Twitch',value:'Twitch'},
                {element:'Twitter',value:'Twitter'},
            ]}/> 
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('should show list', () => {
    const onPress = jest.fn();
    const wrapper = shallow(<Dropdown 
        title="--selection--" 
        onPress={onPress}
        items={[
            {element:'Facebook',value:'Facebook'},
            {element:'Twitch',value:'Twitch'},
            {element:'Twitter',value:'Twitter'},
      ]}/>);
    wrapper.find(TouchableOpacity).first().props().onPress();
    expect(wrapper).toMatchSnapshot();
  });
  it('should Change Selection', () => {
    const onChange = jest.fn();
    const wrapper = shallow(<Dropdown 
        title="--selection--" 
        onChange={onChange}
        items={[
            {element:'Facebook',value:'Facebook'},
            {element:'Twitch',value:'Twitch'},
            {element:'Twitter',value:'Twitter'},
      ]}/>);

    wrapper.find(TouchableOpacity).at(1).props().onPress();
    expect(onChange).toHaveBeenCalledTimes(1);
  });
  it('should not change selection when disabled', () => {
    const onChange = jest.fn();
    const wrapper = shallow(<Dropdown 
                              disabled
                              onChange={onChange}
                              title="--selection disabled--" 
                              items={[
                                  {element:'Facebook',value:'Facebook'},
                                  {element:'Twitch',value:'Twitch'},
                                  {element:'Twitter',value:'Twitter'},
                              ]}/>
                             );
    wrapper.simulate('press');
    expect(onChange).not.toHaveBeenCalled();
  });
  it('should display dark theme', () => {
    const component = shallow(<Dropdown 
                theme="dark"
                title="--selection--" 
                items={[
                    {element:'Facebook',value:'Facebook'},
                    {element:'Twitch',value:'Twitch'},
                    {element:'Twitter',value:'Twitter'},
                ]}/>
      );
    expect(component.length).toBe(1);
    expect(toJson(component)).toMatchSnapshot();
  });
});
