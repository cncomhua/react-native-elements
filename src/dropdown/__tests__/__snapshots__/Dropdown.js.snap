// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown Component renders correctly 1`] = `
<View
  style={
    Object {
      "borderColor": "black",
      "borderWidth": 1,
      "opacity": 1,
      "zIndex": 1,
    }
  }
>
  <View
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "borderBottomWidth": 0,
        "borderColor": "black",
        "borderRadius": 0,
        "borderWidth": 0,
        "flexDirection": "row",
        "justifyContent": "space-between",
        "opacity": 1,
        "overflow": "hidden",
      }
    }
  >
    <View
      style={
        Object {
          "alignSelf": "center",
        }
      }
    >
      <Text
        style={
          Object {
            "color": "black",
            "fontSize": 15,
            "paddingHorizontal": 2,
            "paddingVertical": 2,
          }
        }
      >
        --selection--
      </Text>
    </View>
    <View
      style={
        Object {
          "alignSelf": "center",
        }
      }
    >
      <View
        style={
          Object {
            "overflow": "hidden",
          }
        }
      >
        <View>
          <View
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "transparent",
                "justifyContent": "center",
              }
            }
          >
            <Text
              allowFontScaling={false}
              style={
                Array [
                  Object {
                    "color": "black",
                    "fontSize": 20,
                  },
                  Object {
                    "backgroundColor": "transparent",
                  },
                  Object {
                    "fontFamily": "Feather",
                    "fontStyle": "normal",
                    "fontWeight": "normal",
                  },
                  Object {},
                ]
              }
              testID="iconIcon"
            >
              ï„®
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
  <RCTScrollView
    scrollEventThrottle={0.0001}
    style={
      Object {
        "height": 0,
        "overflow": "hidden",
      }
    }
  >
    <View>
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "flexDirection": "row",
            "justifyContent": "space-between",
            "opacity": 1,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "black",
              "fontSize": 15,
              "paddingHorizontal": 2,
              "paddingVertical": 2,
            }
          }
        >
          Facebook
        </Text>
      </View>
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "flexDirection": "row",
            "justifyContent": "space-between",
            "opacity": 1,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "black",
              "fontSize": 15,
              "paddingHorizontal": 2,
              "paddingVertical": 2,
            }
          }
        >
          Twitch
        </Text>
      </View>
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "flexDirection": "row",
            "justifyContent": "space-between",
            "opacity": 1,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "black",
              "fontSize": 15,
              "paddingHorizontal": 2,
              "paddingVertical": 2,
            }
          }
        >
          Twitter
        </Text>
      </View>
    </View>
  </RCTScrollView>
</View>
`;

exports[`Dropdown Component should display dark theme 1`] = `
<View
  style={
    Object {
      "borderColor": "white",
      "borderWidth": 1,
      "opacity": 1,
      "zIndex": 1,
    }
  }
>
  <ForwardRef
    disabled={false}
    onPress={[Function]}
    style={
      Object {
        "borderBottomWidth": 0,
        "borderColor": "white",
        "borderRadius": 0,
        "borderWidth": 0,
        "flexDirection": "row",
        "justifyContent": "space-between",
        "overflow": "hidden",
      }
    }
  >
    <View
      style={
        Object {
          "alignSelf": "center",
        }
      }
    >
      <Text
        style={
          Object {
            "color": "white",
            "fontSize": 15,
            "paddingHorizontal": 2,
            "paddingVertical": 2,
          }
        }
      >
        --selection--
      </Text>
    </View>
    <View
      style={
        Object {
          "alignSelf": "center",
        }
      }
    >
      <Themed.Icon
        color="white"
        name="chevron-down"
        size={20}
        type="feather"
      />
    </View>
  </ForwardRef>
  <ForwardRef(AnimatedComponentWrapper)
    style={
      Object {
        "height": 0,
        "overflow": "hidden",
      }
    }
  >
    <ForwardRef
      key="0"
      onPress={[Function]}
      style={
        Object {
          "flexDirection": "row",
          "justifyContent": "space-between",
        }
      }
    >
      <Text
        style={
          Object {
            "color": "white",
            "fontSize": 15,
            "paddingHorizontal": 2,
            "paddingVertical": 2,
          }
        }
      >
        Facebook
      </Text>
    </ForwardRef>
    <ForwardRef
      key="1"
      onPress={[Function]}
      style={
        Object {
          "flexDirection": "row",
          "justifyContent": "space-between",
        }
      }
    >
      <Text
        style={
          Object {
            "color": "white",
            "fontSize": 15,
            "paddingHorizontal": 2,
            "paddingVertical": 2,
          }
        }
      >
        Twitch
      </Text>
    </ForwardRef>
    <ForwardRef
      key="2"
      onPress={[Function]}
      style={
        Object {
          "flexDirection": "row",
          "justifyContent": "space-between",
        }
      }
    >
      <Text
        style={
          Object {
            "color": "white",
            "fontSize": 15,
            "paddingHorizontal": 2,
            "paddingVertical": 2,
          }
        }
      >
        Twitter
      </Text>
    </ForwardRef>
  </ForwardRef(AnimatedComponentWrapper)>
</View>
`;

exports[`Dropdown Component should show list 1`] = `ShallowWrapper {}`;
